src/WebJobs.Extensions.DurableTask/Listener/TaskEntityShim.cs

```diff
@@ -237,7 +237,7 @@ public override async Task<string> Execute(OrchestrationContext innerContext, st
                         this.context.Name,
                         this.context.InstanceId,
                         this.entityTraceInfo.TraceFlags,
-                        this.context.InternalError.ToString());
+                        this.context.InternalError.SourceException);
```

Comment:
not immediately sure if this is right

Reply:
Can we add a test or run an E2E test to validate this change?

Reply:
Do the tests that you added cover this scenario?

Reply:
Not quite, but I think I can explain why this makes sense and why it may be odd to write a test for this. I'm gonna have to write a small wall of text, sorry about that!

First - some background.

So this `InternalError` variable is of type ExceptionDispatchInfo, which is a built-in .NET type described here: https://learn.microsoft.com/en-us/dotnet/api/system.runtime.exceptionservices.exceptiondispatchinfo?view=net-8.0. In short, and as described in the docs, you can use this class to _capture_ and exception and re-throw it later without losing the original stack trace.

Here's an example of that capture and rethrow behavior: https://github.com/Azure/azure-functions-durable-extension/blob/829777ae7efb3dbbbab396f4bd4d0f76bc3c0955/src/WebJobs.Extensions.DurableTask/EventGridLifeCycleNotificationHelper.cs#L444 . The benefit here is the **throw** will preserve the original stack trace, instead of generating a new one at the current throw location.

As the documentation also states, the **SourceException** property allows you to obtain the original exception captured within.

Second - some experiments

I wrote the following Console app to demonstrate the logging behavior of `ExceptionDispatchInfo`, see below:

```C#
// See https://aka.ms/new-console-template for more information
using System.Runtime.ExceptionServices;

var rawException = new Exception("to be captured");
var wrappedException = ExceptionDispatchInfo.Capture(rawException);

var rawException2 = wrappedException.SourceException;
Console.WriteLine($"rawException: {rawException.ToString()}");
Console.WriteLine($"wrappedException: {wrappedException.ToString()}");
Console.WriteLine($"rawException2: {rawException2.ToString()}");
```
 
The output is the following:

```
rawException: System.Exception: to be captured
wrappedException: System.Runtime.ExceptionServices.ExceptionDispatchInfo
rawException2: System.Exception: to be captured
```

As you can see, logging the original exception, and the **SourceException** yields the same result. But logging the `<ExceptionDispatchInfo>.toString()` yields simply " System.Runtime.ExceptionServices.ExceptionDispatchInfo".

_So - what does this mean?_ It means that, in the code **today**, we're not logging the exception at all in **EntityBatchFailed**, we're just logging the string "System.Runtime.ExceptionServices.ExceptionDispatchInfo".

You can confirm this on Kusto, by running the following query:

```kql
All("DurableFunctionsEvents")
| where PreciseTimeStamp between(ago(3d).. now()) // p >= datetime(2024-06-19 16:24:00) and PreciseTimeStamp <= datetime(2024-06-19 16:25:00)
| where TaskName =~ "EntityBatchFailed" and Details contains "System.Runtime.ExceptionServices.ExceptionDispatchInfo"
| project TaskName, Details
| take 5
``` 

and you'll see

TaskName | Details
-- | --
EntityBatchFailed | System.Runtime.ExceptionServices.ExceptionDispatchInfo
EntityBatchFailed | System.Runtime.ExceptionServices.ExceptionDispatchInfo
EntityBatchFailed | System.Runtime.ExceptionServices.ExceptionDispatchInfo
EntityBatchFailed | System.Runtime.ExceptionServices.ExceptionDispatchInfo
EntityBatchFailed | System.Runtime.ExceptionServices.ExceptionDispatchInfo
```

**So what is this change doing then?**

Well, I am **assuming** that we meant to log the exception, not this "ExceptionDispatchInfo" constant, so I believe there's a bug. The fix I'm implementing here is to log the actual exception, via **SourceException**, and to use the sanitization utilities I've implemented to sanitize it. So yes, this is a breaking change, but I'd argue it's more akin to a bug fix - we were previously logging a constant string, instead of exception data that can be used for debugging.

Hope this makes sense, let me know what you think!

Reply:
This makes sense. Thanks for the detailed explanation!